### DISCUSSION TITLE
[2022/03/01] Frontend Night

### DISCUSSION THREAD

**Discussion by jjenko** [wantedly/frontend-night#194] [Archive]
## Policy

- 資料の準備をがんばらない
- 参加・途中退席は自由
- スキップしない

話したい・聞きたいネタを書いてく ✏️

ハッシュタグ [#frontend_night](https://twitter.com/hashtag/frontend_night?src=hash)

Prev: https://github.com/wantedly/frontend-night/discussions/193 

 _Created from https://github.com/wantedly/frontend-night/discussions/179 by [issue-creator](https://github.com/rerost/issue-creator)_

**Comment by kobayang** [comment]
## [Node.jsのMySQLパッケージにおけるエスケープ処理だけでは防げない「隠れた」SQLインジェクション](https://blog.flatt.tech/entry/node_mysql_sqlinjection)

- object を SQL に渡していたら予期せぬ SQLインジェクションを起こすことがあるから気をつけようという話。

**Comment by kobayang** [comment]
## [「推測不可能性」だけで安心？セキュアなURL生成について考える](https://blog.flatt.tech/entry/url_issue_security)

> 他ユーザーもしくは不特定多数と情報を共有するためにURLを発行する機能をWebアプリケーション上で実装する際、用途に合わせた適切なアクセス制御を実施することです。

- アクセス制御の例
  - 厳格なアクセス制御（e.g. 予約投稿)
  - 一時的にアクセスを許可する（e.g. 招待URL)
    - 別サービス（記事ではAWS署名付きURL）を使うことも考えられる

そもそも予測可能なURLはスクレイピングされやすいからよくないよね、という話は別でありそう（最近議論になってた）

**Comment by chloe463** [comment]
Tailwind 食わず嫌いしてたけど、週末ちょっと遊んでみたので小並感を共有

- 意外と class name 分からないわけではないなと思った。 
	- padding-top → pt, margin-bottom → mb とか
-  `mt-4` とかで margin-top 取れるが、末尾の 4 は preset されてる数値に変換されるから、きっちり spacing の unit が定義されてるデザインシステムとかと相性が良いかもしれない？
	- default 単位が `rem` で定義されているが tailwind.config.js でカスタマイズできるので、例えば `1: 4px, 2: 8px,...` みたいに定義しておけば Wantedly のデザインシステムにもマッチはしそう
	- これによって変な奇数を使ってしまったり、4でも8でも割り切れない偶数を使ってしまったりしない
	- 一応 `w-[7px]` みたいな書き方で抜け道がないことはない
-  prop によって style を切り替えるときがちょっと面倒かもしれない。
	- styled-components だと下記でできるけど
	```
	<Hoge $loading={isLoading} />
	styled.div<{ $loading: boolean }>`
		color: ${({ $loading }) => $loading ? gray : white};
	`
	```
	- Tailwind でやろうとするとこんな感じ？(雰囲気コーディングだから間違ってそう)
	```
	<div className=`pd-4 ${loading ? text-gray-800 : text-white}`
	```
	-  classnames と組み合わせろってことかな
-  当たり前かもだけど `className=`  につづく文字列が長くなりがちなのでパッと見で見にくい、理解しにくい問題があるかもしれない
	- レビュアーも慣れていないといけない

**Reply by kobayang** [reply]
最近自分が思ってることとしては、 個別のコンポーネントのスタイルをデザインシステムで担保しつつ、レイアウトや余白部分は tailwinds に乗っかるのが、一番開発生産性が良いのではないかと思っている。

styled-components 好きなんだけど、 毎回コンポーネントに命名する必要があるのと、あと、`attr` で型推論が効かないのが地味に痛い...

スタイルに条件が入る場合は、`classNames` で宣言的に書く、のが良いのかなぁ。

**Reply by kobayang** [reply]
tailwind、レスポンシブ対応しなくちゃいけなくなった時にどうなるかはちょっと気になる

**Reply by chloe463** [reply]
- styled-components で最近ちょっとむずかしいなと思うのは、ちゃんと semantic html を意識して書きたいと思ったときに、jsx だけを見てそれが html 的に正しそうかが分からないところですね...
	- styled-components の命名規則に html tag 名まで入れたらいいのかもしれない？ (TitleH2 とか DescriptionP とか？)
-  responsive 対応は `md:` とか `lg:` とか prefix をつけてできるっぽいですね → https://tailwindcss.com/docs/responsive-design

**Reply by kobayang** [reply]
>  prefix をつけてできるっぽいですね

prefix つけるのは知ってるんですが、それをした時に複雑になりすぎてメンテナンス性が下がるのでは？という懸念があるんですよね
（メディアクエリをメンテナブルに実装するのが本質的にむずいと言う話はありそう。）
